fedimint-atomic-broadcast/src/session.rs

```diff
@@ -0,0 +1,242 @@
+    // attack subsides.
+    config.max_round = MAX_ROUND;
+    config.delay_config.unit_creation_delay = std::sync::Arc::new(|round_index| {
+        let delay = if round_index == 0 {
+            0.0
+        } else if round_index < EXPONETIAL_SLOWDOWN_OFFSET {
+            ROUND_DELAY
+        } else {
+            ROUND_DELAY * BASE.powf((round_index - EXPONETIAL_SLOWDOWN_OFFSET) as f64)
+        };
+
+        Duration::from_millis(delay.round() as u64)
+    });
+
+    // the number of units ordered in a single aleph session is bounded
+    let (unit_data_sender, unit_data_receiver) = async_channel::unbounded();
+    let (signature_sender, signature_receiver) = watch::channel(None);
+    let (terminator_sender, terminator_receiver) = futures::channel::oneshot::channel();
+
+    let aleph_handle = tokio::spawn(aleph_bft::run_session(
+        config,
+        aleph_bft::LocalIO::new(
+            DataProvider::new(keychain.clone(), item_receiver, signature_receiver),
+            FinalizationHandler::new(unit_data_sender),
+            backup_saver,
+            backup_loader,
+        ),
+        Network::new(network_data_receiver, outgoing_message_sender.clone()),
+        keychain.clone(),
+        Spawner::new(),
+        aleph_bft::Terminator::create_root(terminator_receiver, "Terminator"),
+    ));
+
+    // we periodically request the signed block corresponding to the current session
+    // to recover in case we have been left behind by our peers
+    let block_request_handle = tokio::spawn(async move {
+        for peer_id in (0..keychain.peer_count)
+            .map(|peer_index| to_peer_id(peer_index.into()))
+            .cycle()
+        {
+            if outgoing_message_sender
+                .send((Message::BlockRequest(index), Recipient::Peer(peer_id)))
+                .await
+                .is_err()
+            {
+                break;
+            }
+
+            sleep(BLOCK_REQUEST_DELAY).await;
+        }
+    });
+
+    // this is the minimum number of unit data that will be ordered before we reach
+    // the EXPONETIAL_SLOWDOWN_OFFSET even if no malicious peer attaches unit
+    // data
+    let n_batches_per_block = EXPONETIAL_SLOWDOWN_OFFSET * keychain.peer_count / 3;
+    let mut items = vec![];
+    let mut n_batches = 0;
+
+    // we build a block out of the ordered batches until either we have processed
+    // n_batches_per_block blocks or a signed block arrives from our peers
+    while n_batches < n_batches_per_block {
+        tokio::select! {
+            unit_data = unit_data_receiver.recv() => {
+                if let UnitData::Batch(batch_items, signature, node_index) = unit_data? {
+                    let mut engine = sha256::HashEngine::default();
+                    batch_items
+                        .consensus_encode(&mut engine)
+                        .expect("Writing to HashEngine cannot fail");
+                    let hash = sha256::Hash::from_engine(engine);
```

Comment:
You could throw that hash calculation in some standalone function somewhere. I've seen Eric asked about it and there's some bitcoin crate version issue, but one way or another it should be extractable to a function.

Reply:
Done